src/control/lib/netdetect/netdetect.go

```diff
@@ -430,6 +430,59 @@ func getNUMASocketID(topology C.hwloc_topology_t, node C.hwloc_obj_t) (uint, err
 	return 0, nil
 }
 
+// GetNUMASocketIDForPid determines the cpuset and nodeset corresponding to the given pid.
+// It looks for an intersection between the nodeset or cpuset of this pid and the nodeset or cpuset of each
+// NUMA node looking for a match to identify the corresponding NUMA socket ID.
+func GetNUMASocketIDForPid(pid int32) (uint, error) {
+	var i uint
+
+	deviceScanCfg, err := initDeviceScan()
+	if err != nil {
+		return 0, err
+	}
+	defer cleanUp(deviceScanCfg.topology)
+
+	depth := C.hwloc_get_type_depth(deviceScanCfg.topology, C.HWLOC_OBJ_NUMANODE)
+	numObj := uint(C.hwloc_get_nbobjs_by_depth(deviceScanCfg.topology, C.uint(depth)))
+	if numObj == 0 {
+		log.Debugf("NUMA Node data is unavailable.  Using NUMA 0\n")
+		return 0, nil
+	}
+
+	cpuset := C.hwloc_bitmap_alloc()
+	defer C.hwloc_bitmap_free(cpuset)
+	status := C.hwloc_get_proc_cpubind(deviceScanCfg.topology, C.int(pid), cpuset, 0)
+	if status != 0 {
+		log.Debugf("hwloc_get_proc_cpubind failure!")
+		return 0, nil
+	}
+
+	nodeset := C.hwloc_bitmap_alloc()
+	defer C.hwloc_bitmap_free(nodeset)
+	C.hwloc_cpuset_to_nodeset(deviceScanCfg.topology, cpuset, nodeset)
+
+	for i = 0; i < numObj; i++ {
+		numanode := C.hwloc_get_obj_by_depth(deviceScanCfg.topology, C.uint(depth), C.uint(i))
+		if numanode == nil {
+			// We don't want the lack of NUMA information to be an error.
+			// If we get this far and can't access the NUMA topology data,
+			// we will use NUMA ID 0.
+			log.Debugf("NUMA Node data is unavailable.  Using NUMA 0\n")
+			return 0, nil
```

Comment:
Not sure I agree with this design decision. It seems like it would be better to let the caller decide what to do if the numa node isn't available.

Reply:
If the client is not bound, and therefore, no NUMA node information is available, how would the client either recover, or manage to make a valid GetAttachInfo call?  At the moment, we have the requirement that the agent determine the NUMA affinity on the client's behalf because we do not trust the client.  So that rules out having the client pass in a desired NUMA node.  If the client isn't going to pass in a NUMA node, and the agent can't figure it out, it could be treated as an error.  I am fine defining that condition as an error if we want.

Reply:
I'm just pointing out that this is a library function and could be used by callers other than the agent's getAttachInfo functionality. The behavior you have now works for that one use case, but might not be correct for other use cases. Returning an error and letting the caller decide what to do about not being able to determine the correct numa node seems more future-proof.

Reply:
That's fair.  I can make that change for sure.